### DISCUSSION TITLE
certificate problems when application is working as TLS client

### DISCUSSION THREAD

**Discussion by shubhamfuloria** [resiprocate/resiprocate#358] [Q&A] [answered]
When resiprocate tries to refresh the sesssion, we get logs
```
TLS handshake starting (client mode)  
TLS connected  
// then we get an error
Certificate name mismatch: trying to connect to <10.1.1.161> remote cert domain(s) are <example.com, www.example.com>
```
All certificate of client and server are stored correctly as mentioned in [documentation](https://github.com/resiprocate/resiprocate/wiki/Using-TLS). I checked source code of resiprocate and the error is from [resip/stack/ssl/TlsConnection.cxx](https://github.com/resiprocate/resiprocate/blob/master/resip/stack/ssl/TlsConnection.cxx#L214) 's this section

```cpp
      else
      {
         InfoLog( << "TLS handshake starting (client mode)" );
         /* OpenSSL version < 0.9.8f does not support SSL_set_tlsext_host_name() */
#if defined(SSL_set_tlsext_host_name)
            DebugLog ( << "TLS SNI extension in Client Hello: " << who().getTargetDomain());
            SSL_set_tlsext_host_name(mSsl,who().getTargetDomain().c_str()); // set the SNI hostname
#endif
         SSL_set_connect_state(mSsl);
         mTlsState = Handshaking;
      }
```

I added some additional logs in this section, and `who().getTargetDomain()` returns `ip address` of server, instead of domain name, that's why I get this error. Is there an API by which we can set domain name ? How am I suppose to set domain name when resiprocate is working as client in tls handshake ?

**Comment by sgodin** [comment]
As long as you try to reach the server using it's TLS domain name and not it's IP directly, then things should work.  

Since this is happening on a TLS session refresh, perhaps something in the SIP routing headers is using the IP address of the TLS server, instead of it's domain name?   If  the server is passing it's IP in any routing headers, then it must be prepared to handle a TLS connection using the IP directly and so it must present it's IP address in it's server certificate.  

The logs should indicate where the server's IP is coming from (if other than DNS).

**Reply by shubhamfuloria** [reply]
Initally client connects with server and TLS handshake happens with these logs
```
INFO | 20240412-145134.378 |  | RESIP:TRANSPORT | 140011708508288 | Connection.cxx:45 | Connection::Connection: new connection created to who: [ V4 [10.1.1.161:33177](http://10.1.1.161:33177/) TLS flowKey=11 transportKey=1 ], $
INFO | 20240412-145134.378 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:75 | Creating TLS connection for domain [srcp02-rtel.rtel.gr](http://srcp02-rtel.rtel.gr/) [ V4 [10.1.1.161:33177](http://10.1.1.161:33177/) TLS transportKey=1 ] on 11
INFO | 20240412-145134.378 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:80 | mDomain value: [exampl1.com](http://exampl1.com/)
INFO | 20240412-145134.387 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:222 | TLS handshake starting (Server mode)
INFO | 20240412-145134.387 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:242 | TLS connected
INFO | 20240412-145134.398 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:359 | TLS connected
INFO | 20240412-145134.398 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:722 | TLS sessions set up with TLSv1 TLSv1/SSLv3 AES128-SHA
INFO | 20240412-145134.398 |  | RESIP:TRANSPORT | 140011708508288 | ssl/TlsConnection.cxx:395 | TLS handshake done for peer
```
Then after 5 minutes application tries to refreh the session, this time TLS handshake starts in (Client mode)

**Reply by sgodin** [reply, marked as answer]
Since you only included a very small log snippet, it's hard to tell exactly what's happening, but I'll make some assumptions.   

Is it a SIP session refresh that causes this?  Is this a DUM based application or repro?  By default, SIP (and reSIP) doesn't do connection re-use with TCP based transports (like TLS).  ie:  If the client sends resip a request, the response will go down the same connection, however any SIP requests going from the server back to client, will attempt to form a new connection from the server back to client.   And in your case the client probably isn't configured with a certificate that has either the hostname or IP address from it's routing or contact headers. 

If your server is repro based, there are settings you can enable to turn on connection re-use, and repro supports RFC5626 if you enable the correct settings.  If you are building a client application with DUM you have Profile/UserProfile settings to enable this behaviour (see below).  If you are building your own server with DUM then you will need to build in this behaviour (see SipMessage::setDestination and it's use in repro).

It's possible I've made bad assumptions and you are seeing some other issue entirely.

Some useful links:
https://github.com/resiprocate/resiprocate/wiki/DUM-Configuration-and-Profiles
https://github.com/resiprocate/resiprocate/wiki/DUM-Client-Outbound-Support
https://github.com/resiprocate/resiprocate/wiki/Using-Repro (see flow tokens and outbound settings)
http://www.sipspectrum.com/blog/configuring-repro-for-webrtc
https://datatracker.ietf.org/doc/html/rfc5626
